### PR TITLE
Add support for TypeScript in Vue SFCs.

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description

Adds support for TypeScript in Vue SFCs.

Notable changes:
* Usage of `vue-tsc` and addition of some modules that Vue 2 doesn't automatically pull in necessary for template types
* Applied fixes from https://github.com/johnsoncodehk/volar/discussions/592 due to Storybook pulling `@types/react` into the chain causing templates to have the incorrect types applied to them by TypeScript compiler

I chose a simple `VLink` component to be the test candidate.

## Testing Instructions

Checkout this branch and make changes in VLink and run `pnpm types` test compilation.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: src/components/VLink.vue [hunk 1]
```diff
@@ -16,14 +16,18 @@
...
+})
 
 export default defineComponent({
   name: 'VLink',
   props: {
     href: {
       type: String,
-      validator: (v) =>
+      required: false,
+      validator: (v: unknown) =>
```
Comment (sarayourfriend) [inline src/components/VLink.vue:30]: Please note that setting `validator`'s argument explicitly to `unknown` is necessary for prop extraction to work when a `validator` is present. Even though `unknown` is the inferred type. :shrug:

**Review by dhruvkb — APPROVED** [review]
I had expected a lot more “fighting the toolchain” here. This looks so much more pleasant.

**Comment by sarayourfriend** [comment]
@dhruvkb Me too! Getting the `@types/react` to drop was the hardest part, mostly because I resisted using the `types` compiler option. Once I resigned to using it I expected to have to include a long list of packages in that list but luckily only `jest` needed it! :tada: We could even avoid that if we decided to explicitly import `test` and `expect` from the `jest` package.

**Comment by dhruvkb** [comment]
Interesting! I'm really looking forward to the frontend becoming a 100% TypeScript codebase.

**Comment by sarayourfriend** [comment]
Me too :grinning:

**Review by krysal — APPROVED** [review]
Amazing, tested with some types, and works as expected! 🚀

**Pull request merged**